events/ratelimiting/coalescing.go

```diff
@@ -0,0 +1,256 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ratelimiting
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"k8s.io/utils/clock"
+)
+
+// OptionsCoalescing configures a Coalescing RateLimiter.
+type OptionsCoalescing struct {
+	// InitialDelay is the initial delay for the rate limiter. The rate limiter
+	// will not delay events less than the initial delay.
+	// Defaults to 500ms.
+	InitialDelay *time.Duration
+
+	// MaxDelay is the maximum delay for the rate limiter. The rate limiter will
+	// not delay events longer than the max delay.
+	// Defaults to 5s.
+	MaxDelay *time.Duration
+
+	// MaxPendingEvents is the maximum number of events that can pending on a
+	// rate limiter, before it fires an event anyway. Useful to prevent a rate
+	// limiter never firing events in a high throughput scenario.
+	// Defaults to unlimited.
+	MaxPendingEvents *int
+}
+
+// coalescing is a rate limiter that rate limits events. It coalesces events
+// that occur within a rate limiting window.
+type coalescing struct {
+	initialDelay     time.Duration
+	maxDelay         time.Duration
+	maxPendingEvents *int
+
+	pendingEvents int
+	timer         clock.Timer
+	hasTimer      atomic.Bool
+	inputCh       chan struct{}
+	currentDur    time.Duration
+	backoffFactor int
+
+	wg      sync.WaitGroup
+	lock    sync.RWMutex
+	clock   clock.WithTicker
+	running atomic.Bool
+	closeCh chan struct{}
+	closed  atomic.Bool
+}
+
+func NewCoalescing(opts OptionsCoalescing) (RateLimiter, error) {
+	initialDelay := time.Millisecond * 500
+	if opts.InitialDelay != nil {
+		initialDelay = *opts.InitialDelay
+	}
+	if initialDelay < 0 {
+		return nil, errors.New("initial delay must be > 0")
```

Comment:
You say "must be > 0" but the check allows everything ">= 0"

Reply:
Fixed